torch/csrc/api/src/data/samplers/distributed.cpp

```diff
@@ -0,0 +1,158 @@
+#include <torch/data/samplers/distributed.h>
+#include <torch/serialize/archive.h>
+#include <torch/types.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <random>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+DistributedRandomSampler::DistributedRandomSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : DistributedSampler(size, num_replicas, rank, allow_duplicates),
+      begin_index_(0),
+      end_index_(0),
+      sample_index_(0) {
+  populate_indices();
+  // shuffle first time.
+  reset(size_);
+}
+
+optional<std::vector<size_t>> DistributedRandomSampler::next(
+    size_t batch_size) {
+  if (sample_index_ == end_index_) {
+    return nullopt;
+  }
+
+  size_t end = sample_index_ + batch_size;
+  if (end > end_index_) {
+    end = end_index_;
+  }
+
+  auto iter = all_indices_.begin();
+  std::vector<size_t> res(iter + sample_index_, iter + end);
+  sample_index_ = end;
+  return res;
+}
+
+void DistributedRandomSampler::reset(optional<size_t> new_size) {
+  size_t size = new_size.value_or(size_);
+  if (size != size_) {
+    size_ = size;
+    populate_indices();
+  }
+  std::mt19937 rand(epoch_);
+  std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
```

Comment:
It is somewhat weird that if you set the epoch to 1 and then 2 in one process, and to 3 and then 2 in another one, the indices you'd get in the epoch marked as 2 would be different. I think it would be much better to make the shuffling independent of the sequence of previous epochs. Basically, a new `DistributedRandomSampler` with an epoch of 5 should be equivalent to one that had the epoch set to 1, 2, ..., 5. Otherwise you can't e.g. reliably restart from a checkpoint.

Reply:
I see, good point. in python version we create indices in each __iter__ call, so the shuffling is the same based on the epoch. Fixed it.